dash-spv-coinjoin/src/coinjoin_client_session.rs

```diff
@@ -478,62 +485,210 @@ impl CoinJoinClientSession {
+        // Knowing that each CTxIn is at least 148 B big, 400 inputs should take 400 x ~148 B = ~60 kB.
+        // This still leaves more than enough room for another data of typical MakeCollateralAmounts tx.
+        let mut vec_tally = self.mixing_wallet.borrow_mut().select_coins_grouped_by_addresses(false, false, true, 400);
+
+        if vec_tally.is_empty() {
+            println!("[RUST] CoinJoinClientSession::MakeCollateralAmounts -- SelectCoinsGroupedByAddresses can't find any inputs!\n");
+            return false;
+        }
+
+        // Start from the smallest balances first to consume tiny amounts and cleanup UTXO a bit
+        vec_tally.sort_by(|a, b| a.amount.cmp(&b.amount));
+
+        // First try to use only non-denominated funds
+        for item in &vec_tally {
+            if !self.make_collateral_amounts_with_item(item, false) {
+                continue;
+            }
+
+            return true;
+        }
+
+        // There should be at least some denominated funds we should be able to break in pieces to continue mixing
+        for item in &vec_tally {
+            if !self.make_collateral_amounts_with_item(item, true) {
+                continue;
+            }
+
+            return true;
+        }
+
+        // If we got here then something is terribly broken actually
+        println!("[RUST] CoinJoinClientSession::MakeCollateralAmounts -- ERROR: Can't make collaterals!\n");
         return false;
     }
 
+    fn make_collateral_amounts_with_item(&mut self, tally_item: &CompactTallyItem, try_denominated: bool) -> bool {
+        if !self.options.enable_coinjoin {
+            return false;
+        }
+
+        println!("[RUST] CoinJoin make_collateral_amounts_with_item: {:?}", tally_item);
+
+        // Denominated input is always a single one, so we can check its amount directly and return early
+        if !try_denominated && tally_item.input_coins.len() == 1 && CoinJoin::is_denominated_amount(tally_item.amount) {
+            return false;
+        }
+
+        // Skip single inputs that can be used as collaterals already
+        if tally_item.input_coins.len() == 1 && CoinJoin::is_collateral_amount(tally_item.amount) {
+            return false;
+        }
+
+        let mut tx_builder = TransactionBuilder::new(
+            self.mixing_wallet.clone(),
+            self.options.chain_type,
+            tally_item.clone(),
+            false
+        );
+        println!("[RUST] CoinJoin make_collateral_amounts_with_item. Start tx_builder {}", tx_builder);
+        
+        // Skip way too tiny amounts. Smallest we want is minimum collateral amount in a one output tx
+        if !tx_builder.could_add_output(CoinJoin::get_collateral_amount()) {
+            return false;
+        }
+
+        let case; // Just for debug logs
+
+        if tx_builder.could_add_outputs(&[CoinJoin::get_max_collateral_amount(), CoinJoin::get_collateral_amount()].to_vec()) {
```

Comment:
If I'm not hallucinating, then there you can use array notation &[] in "could_add_outputs" rather then &Vec<_> and therefore avoid unnecessary allocations